util/src/journaldb/overlayrecentdb.rs

```diff
@@ -99,34 +103,53 @@ const PADDING : [u8; 10] = [ 0u8; 10 ];
 
 impl OverlayRecentDB {
 	/// Create a new instance.
-	pub fn new(backing: Arc<Database>, col: Option<u32>) -> OverlayRecentDB {
-		let journal_overlay = Arc::new(RwLock::new(OverlayRecentDB::read_overlay(&backing, col)));
+	pub fn new(backing: Arc<Database>, col: Option<u32>, mode: Archive) -> OverlayRecentDB {
+		let overlay_col = match mode {
+			Archive::Off => col,
+			Archive::On(c) => c,
+		};
+
+		let journal_overlay = Arc::new(RwLock::new(OverlayRecentDB::read_overlay(&backing, overlay_col)));
 		OverlayRecentDB {
 			transaction_overlay: MemoryDB::new(),
 			backing: backing,
 			journal_overlay: journal_overlay,
 			column: col,
+			mode: mode,
 		}
 	}
 
 	/// Create a new instance with an anonymous temporary database.
 	#[cfg(test)]
-	pub fn new_temp() -> OverlayRecentDB {
+	pub fn new_temp(mode: Archive) -> OverlayRecentDB {
 		let mut dir = env::temp_dir();
 		dir.push(H32::random().hex());
 		let backing = Arc::new(Database::open_default(dir.to_str().unwrap()).unwrap());
-		Self::new(backing, None)
+		Self::new(backing, None, mode)
 	}
 
 	#[cfg(test)]
 	fn can_reconstruct_refs(&self) -> bool {
-		let reconstructed = Self::read_overlay(&self.backing, self.column);
+		let overlay_col = match self.mode {
+			Archive::Off => self.column,
+			Archive::On(c) => c,
+		};
+
+		let reconstructed = Self::read_overlay(&self.backing, overlay_col);
 		let journal_overlay = self.journal_overlay.read();
 		*journal_overlay == reconstructed
 	}
 
 	fn payload(&self, key: &H256) -> Option<Bytes> {
-		self.backing.get(self.column, key).expect("Low-level database error. Some issue with your hard disk?")
+		match self.backing.get(self.column, key).expect("Low-level database error. Some issue with your hard disk?") {
```

Comment:
By changing `self.mode` to `Option` (on this level (or perhaps `Into`)) you can get a bit more concise code:

```rust
let val = self.backing.get(self.column, key).expect("...");
val.or_else(|| {
  self.archive_column.map(|col| self.backing.get(col, key).expect(".."))
})
```
or
```rust
self.get_backing(self.column, key).or_else(|| {
  self.archive_column.map(|col| self.get_backing(col, key))
})
 ``